crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -464,14 +483,23 @@ impl RoomEventCacheInner {
         let sender = Sender::new(32);
 
         Self {
-            room,
             events: RwLock::new(RoomEvents::default()),
             sender,
-            pagination_lock: Default::default(),
+            pagination: RoomPaginationData {
+                paginator: Paginator::new(Box::new(room)),
+                waited_for_initial_prev_token: Mutex::new(false),
+            },
             pagination_token_notifier: Default::default(),
         }
     }
 
+    async fn clear(&self, room_events: &mut RwLockWriteGuard<'_, RoomEvents>) {
```

Comment:
Why passing the `room_events`? What's the benefits of this new approach?

Reply:
Initially, I took the `room_events` lock within this function; but when this is called in `replace_all_events_by`, we're already holding onto the lock, and we *really* want to keep it throughout the full function's body, to avoid a bad race when two callers racily call `replace_all_events_by`.